### ISSUE TITLE
Opt and delimited_list causing parse actions not to be called

### ISSUE THREAD

**Issue by elektito** [pyparsing/pyparsing#408]
I am having an issue where one of the parse actions is not being called in certain situations. I managed to minimize the situation to this code snippet:

```python
from pyparsing import Keyword, Word, Opt, alphas, delimited_list

keyword = Keyword('foobar')
untyped_identifier =  ~keyword + Word(alphas)
dotted_vars = delimited_list(untyped_identifier, delim='.')
lvalue = Opt(dotted_vars)
stmt = delimited_list(lvalue)

def parse_identifier(toks):
    print('YAY!', toks)
untyped_identifier.set_parse_action(parse_identifier)

print(dotted_vars.parse_string('B.C'))
```

If I comment out the `stmt=...` line, the parse action is called (twice). If I change the last rule to `stmt = delimited_list(dotted_vars)`, then the parse action is called once (just for "B").

I'm using pyparsing 3.0.9 installed from PyPI.

**Comment by ptmcg** [comment]
Yes, delimitedList does some "stuff" with its embedded expression, changed as part of fixing other bugs. Will take another run at resolving this to retain prior behavior while keeping other bugfixes...  Thanks for reporting.

**Comment by ptmcg** [comment]
Sorry for the delay. The problem is that delimitedList calls streamline() on the expression in order to create a valid label.  streamline() will merge some expressions if they have no special names or parse actions (so that `exprA + exprB + exprC` doesn't display as `(exprA + exprB) + exprC`.

Your code works if you attach the parse action right after defining `untyped_identifier`:

```python
untyped_identifier =  ~keyword + Word(alphas)
def parse_identifier(toks):
    print('YAY!', toks)
untyped_identifier.set_parse_action(parse_identifier)
```

That's not a solution, just an explanation as to where the problem is.

I either have to revert the naming simplifcation for delimited_list expressions, or implement a recursive name generator that won't change the original expression or any sub-expressions (such as an expr inside an Opt).

I'll probably just revert the streamline() call for now, and work on getting a better string representation in a future release.

**Comment by ptmcg** [comment]
Well, it turns out that recursive name generator wasn't that awful after all. So this fix will be in 3.0.10, coming out very soon.

**Comment by ptmcg** [comment]
delimitedList has been reimplemented in the class DelimitedList, and no longer exhibits this problem. delimitedList and delimited_list both point to that class as compatibility synonyms, but will be deprecated in 3.2 and removed in a released to be determined later.

Released in 3.1.0a1
